### PR TITLE
Dominating Set

### PR DESCRIPTION
**Pull request by SohamTamba**
Greedy heuristics for Dominating Set.

We really should consider changing the organisation of these heuristics for NP-Complete optimisation problems (`vertex_cover`, `independent_set`, `dominating_set`).

cc: @somil55  @sbromberger

### CODE REVIEW

**Comment by sbromberger** [comment]
> We really should consider changing the organisation of these heuristics for NP-Complete optimisation problems.

What do you have in mind?

**Comment by SohamTamba** [comment]
I have a few ideas but they all have flaws. Maybe we can make a combination of them

1) Optimisation: Too broad, shortest paths also comes under this category
2) Vertex Set: Too broad
3) NP-Hard/Heuristic: Not very pleasant

How about Optimal Vertex Set?

**Comment by simonschoelly** [comment]
What about calling it "Combinatorics" ?
If you call it NP-Hard/Heuristic, you always must find a proof forthat before you can put your problem there :)

**Comment by SohamTamba** [comment]
@sbromberger  what do you think?

**Comment by sbromberger** [comment]
> @sbromberger what do you think?

I don't have a well-formed opinion on the naming, but if this PR is ready for review, I'm game :) I'll probably want @jpfairbanks or @simonschoelly to review as well.

File: src/dominatingset/degree_dom_set.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    @inbounds if !dominated[v]
+        dominated[v] = true
+        if !in_dom_set[v] 
+            degree_queue[v] -= 1
+        end
+        @inbounds @simd for u in neighbors(g, v)
+            if !in_dom_set[u] 
+                degree_queue[u] -= 1
+            end
```
Comment (simonschoelly) [inline src/dominatingset/degree_dom_set.jl]: I'm surprised, that you can use an if-statement inside a `@simd` loop, I thought the compiler would complain. Still, consider the branchless version
```
degree_queue[u] -= ifelse(in_dom_set[u], 0, 1)
```
This works with self loops, right?

File: test/parallel/dominatingset/minimal_dom_set.jl [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    g5 = PathGraph(4)
+    for parallel in [:threads, :distributed]
+        for g in testgraphs(g5)
+            d = @inferred(Parallel.dominating_set(g, 4, MinimalDominatingSet(); parallel=parallel))
+            sort!(d)
+            @test (d == [1, 2, 4] || d == [1, 3] || d == [1, 4] || d == [2, 3] || d == [2, 4])
+        end
+    end
+end
```
Comment (simonschoelly) [inline test/parallel/dominatingset/minimal_dom_set.jl]: Consider adding test cases for:
* The empty graph
* A single vertex
* A single vertex with a self-loop
* A larger graph with a self-loop

File: src/dominatingset/minimal_dom_set.jl [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+dominating property. If no, the vertex is removed from the dominating set.
+
+### Performance
+Runtime: O(|V|+|E|)
+Memory: O(|V|)
+"""    
+function dominating_set(
+    g::AbstractGraph{T},
+    alg::MinimalDominatingSet
+    ) where T <: Integer
```
[the remark this answers is not in the document]
Comment (jpfairbanks) [reply src/dominatingset/minimal_dom_set.jl]: yes `seed=0` as a kwargs is good.

**Comment by jpfairbanks** [comment]
I agree with @simonschoelly's comments.

File: src/dominatingset/minimal_dom_set.jl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+function dominating_set(
+    g::AbstractGraph{T},
+    alg::MinimalDominatingSet
+    ) where T <: Integer 
+
+    nvg = nv(g)  
+    in_dom_set = trues(nvg)  
+    dom_degree = degree(g)
+    for v in shuffle(vertices(g))
+    	(dom_degree[v] == 0) && continue #It is not adjacent to any dominating vertex
```
[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/dominatingset/minimal_dom_set.jl]: Yeah I didn't consider self-loops while writing this. :/

File: src/dominatingset/minimal_dom_set.jl [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    	
+        safe_to_remove = true
+    	for u in neighbors(g, v)
+        	@inbounds if !in_dom_set[u] && dom_degree[u] <= 1
+        		safe_to_remove = false
+        		break
+        	end
+        end
```
[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/dominatingset/minimal_dom_set.jl]: Yup

`findfirst` to be more precise

File: src/dominatingset/degree_dom_set.jl (hunk 1 above)
Comment (SohamTamba) [reply src/dominatingset/degree_dom_set.jl]: I was thinking the same but `@simd` was used in similar situations for `gdistances!`.

File: src/dominatingset/minimal_dom_set.jl [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        safe_to_remove || continue
+        
+        in_dom_set[v] = false
+        dom_degree[neighbors(g, v)] .-= 1
+    end
+
+    return [v for v in vertices(g) if in_dom_set[v]]
```
[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/dominatingset/minimal_dom_set.jl]: Is there any function `findall!` in base that has the same effect as `findall` but reuses memory?

File: test/parallel/dominatingset/minimal_dom_set.jl (hunk 2 above)
Comment (SohamTamba) [reply test/parallel/dominatingset/minimal_dom_set.jl]: Done

File: src/dominatingset/minimal_dom_set.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/dominatingset/minimal_dom_set.jl]: Sure.

I'll use the same format as [randgraphs](https://github.com/SohamTamba/LightGraphs.jl/blob/bit_BFS/src/SimpleGraphs/generators/randgraphs.jl).

File: src/dominatingset/degree_dom_set.jl (hunk 1 above)
Comment (SohamTamba) [reply src/dominatingset/degree_dom_set.jl]: Also the code you gave won't work since the key `u` may not exist in the queue

Comment (simonschoelly) [reply src/dominatingset/degree_dom_set.jl]: Yeah, you are right, i though it was an array.

File: src/dominatingset/minimal_dom_set.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/dominatingset/minimal_dom_set.jl]: Why do you want to use `findall` here? Wouldn't it be easier to loop over the vector once to calculate how much memory to allocate, then allocate a new vector and then loop over the vector once more  to write the indices to the new vector?

[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/dominatingset/minimal_dom_set.jl]: I mean create a function `findall!(x, A, B)` which writes `findall(x, A)` into `B` (assumes `B` is sufficiently large)

**Comment by SohamTamba** [comment]
@sbromberger I think Travis must be restarted.

Please take a look

**Comment by SohamTamba** [comment]
I think this is ready

**Comment by simonschoelly** [comment]
Isn't your algorithm the one described here: http://ac.informatik.uni-freiburg.de/teaching/ss_12/netalg/lectures/chapter7.pdf ? Then it has an approximation factor.

**Comment by SohamTamba** [comment]
Yeah you're right, the approximation factor is `ln(maximum(degree(g)))+2`

**Pull request merged**